src/code_version.erl

```diff
@@ -0,0 +1,234 @@
+%%----------------------------------------------------------------------------
+load_code(Module, Code) ->
+    unload(Module),
+    case code:load_binary(Module, "loaded by rabbit_common", Code) of
+        {module, _} ->
+            ok;
+        {error, Reason} ->
+            throw({cannot_load, Module, Reason})
+    end.
+
+unload(Module) ->
+    code:soft_purge(Module),
+    code:delete(Module).
+
+compile_forms(Forms) ->
+    case compile:forms(Forms, [debug_info]) of
+        {ok, _ModName, Code} ->
+            Code;
+        {ok, _ModName, Code, _Warnings} ->
+            Code;
+        Error ->
+            throw({cannot_compile_forms, Error})
+    end.
+
+get_abs_code(Module) ->
+    get_forms(get_object_code(Module)).
+
+get_object_code(Module) ->
+    case code:get_object_code(Module) of
+        {_Mod, Code, _File} ->
+            Code;
+        error ->
+            throw({not_found, Module})
+    end.
+
+get_forms(Code) ->
+    case beam_lib:chunks(Code, [abstract_code]) of
+        {ok, {_, [{abstract_code, {raw_abstract_v1, Forms}}]}} ->
+            Forms;
+        {ok, {Module, [{abstract_code, no_abstract_code}]}} ->
+            throw({no_abstract_code, Module});
+        {error, beam_lib, Reason} ->
+            throw({no_abstract_code, Reason})
+    end.
+
+get_otp_version() ->
+    Version = erlang:system_info(otp_release),
+    case re:run(Version, "^[0-9][0-9]", [{capture, first, list}]) of
+        {match, [V]} ->
+            list_to_integer(V);
+        _ ->
+            %% Could be anything below R17, we are not interested
+            0
+    end.
+
+get_original_pairs(VersionSupport) ->
+    [{Orig, Arity} || {Orig, _Pre, _Post, Arity} <- VersionSupport].
+
+get_delete_pairs(true, VersionSupport) ->
+    [{Pre, Arity} || {_Orig, Pre, _Post, Arity} <- VersionSupport];
+get_delete_pairs(false, VersionSupport) ->
+    [{Post, Arity} || {_Orig, _Pre, Post, Arity} <- VersionSupport].
+
+get_rename_pairs(true, VersionSupport) ->
+    [{Post, Arity} || {_Orig, _Pre, Post, Arity} <- VersionSupport];
+get_rename_pairs(false, VersionSupport) ->
+    [{Pre, Arity} || {_Orig, Pre, _Post, Arity} <- VersionSupport].
+
+%% Pairs of {Renamed, OriginalName} functions
+get_name_pairs(true, VersionSupport) ->
+    [{Post, Orig} || {Orig, _Pre, Post, _Arity} <- VersionSupport];
+get_name_pairs(false, VersionSupport) ->
+    [{Pre, Orig} || {Orig, Pre, _Post, _Arity} <- VersionSupport].
+
+delete_abstract_functions(ToDelete) ->
+    fun(Tree, Function) ->
+            case lists:member(Function, ToDelete) of
+                true ->
+                    erl_syntax:comment(["Deleted unused function"]);
+                false ->
+                    Tree
+            end
+    end.
+
+rename_abstract_functions(ToRename, ToName) ->
+    fun(Tree, {Name, _Arity} = Function) ->
+            case lists:member(Function, ToRename) of
+                true ->
+                    FunctionName = proplists:get_value(Name, ToName),
```

Comment:
You ignore the arity of the function about to be renamed. `ToName`, returned by `get_name_pairs/2`, looks like:
```erlang
ToName=[{monotonic_time_post_18,monotonic_time}, % Comes from {monotonic_time, monotonic_time_pre_18, monotonic_time_post_18, 0}
        {monotonic_time_post_18,monotonic_time}, % Comes from {monotonic_time, monotonic_time_pre_18, monotonic_time_post_18, 1}
        % ...
```

It doesn't contain the arity information either, but this is important: here you pick the new name from the first matching entry, ie. always the `{monotonic_time, monotonic_time_pre_18, monotonic_time_post_18, 0}` attribute, even if the new name is different for the `/1` function.